### PR TITLE
Implement linter tags file compatible lintr 3.0.0

### PR DESCRIPTION
**Pull request by m-kolomanski**
Issue #102

## Description
Added *linters.csv* file in /inst/lintr/ with linters and tags, compatiblie with linter 3.0.0. and `lintr::available_linters()`, `lintr::available_tags()` functions.

Added unit tests for this feature, although those need to be run with `devtools::check()` as the mentioned functions expect the package to be installed. Otherwise, related tests will be skipped. 

## Definition of Done
- [x] The change is thoroughly documented.
- [x] The CI passes (`R CMD check`, linter, unit tests, spelling).
- [x] Any generated files have been updated (e.g. `.Rd` files with `roxygen2::roxygenise()`)

### CODE REVIEW

File: tests/testthat/test-linter_tags.R [hunk 1]
```diff
@@ -0,0 +1,36 @@
...
+#' functions can read the csv file with linter definitions.
+#'
+#' lintr expects the package to be already installed. Those tests need to be run
+#' using devtools::check(), otherwise will be skipped.
+#'
+#' Manual checks can also be performed by installing the package and setting the
+#' `linters_csv_location` variable to "./inst/lintr/linters.csv", then running
+#' tests in the console.
+#'
+linters_csv_location <- "../../box.linters/lintr/linters.csv"
```
Comment (radbasa) [inline tests/testthat/test-linter_tags.R:12]: How about checking if it is running under a R CMD check? Refer to https://testthat.r-lib.org/reference/is_testing.html.

Also, take a look at `{rprojroot}` functions to determine the root path for the two different situations.

**Comment by radbasa** [comment]
It might be worth adding `@evalRd rd_tags("...")` to the functions.

https://lintr.r-lib.org/articles/creating_linters.html#writing-the-linter

**Comment by m-kolomanski** [comment]
> It might be worth adding `@evalRd rd_tags("<function_name>")` to the roxygen blocks.
> 
> https://lintr.r-lib.org/articles/creating_linters.html#writing-the-linter

rd_tags is not exported from `lintr` package, therefore this would require creating ./R/linters_tags.R helper file with this function definition. 

Since the `rd_tags` function from `lintr` package generates tags with links that forward to doc pages for specific tags, this would need to be followed with `rd_linters` function and documentation file for each of the tags, with list of linters that have that particular tag assigned.

Alternatively, a different schema could be created for this usecase (like section with list of tags, without links to further doc).

What do you think would be the best for this case?

For details, see [rd_tags()](https://github.com/r-lib/lintr/blob/main/R/linter_tags.R#L145), [rd_linters()](https://github.com/r-lib/lintr/blob/main/R/linter_tags.R#L164), [linter_tags_docs.R](https://github.com/r-lib/lintr/blob/main/R/linter_tag_docs.R) file and generated [example doc](https://github.com/r-lib/lintr/blob/main/man/best_practices_linters.Rd) for `lintr`.

**Comment by radbasa** [comment]
@m-kolomanski Try looking at the following lines of code to use non-exported functions from packages:

https://github.com/Appsilon/box.linters/blob/e5658e433aa3b724650fd308fc162866f731702c/R/get_box_module_exports.R#L21-L24

Maybe, it will work. Maybe, it won't.

File: tests/testthat/test-linter_tags.R (hunk 1 above)
Comment (m-kolomanski) [reply tests/testthat/test-linter_tags.R:12]: Thanks for the pointers, the following could be done to clean up the code a little:

```
linters_csv_location <- file.path(
  rprojroot::find_root_file(criterion = rprojroot::is_r_package),
  "lintr/linters.csv"
)
```

and later in the tests:

```
skip_if(!is_checking())
```

But I don't think I follow the logic with switching files? To my understanding, changing the `linters_csv_location` based on execution context does not matter, as it is the `lintr::available_tags()` function that is failing to read the file when running `devtools::test()`, and we cannot change the bahaviour of that function. Am I missing or misunderstanding something?

**Comment by m-kolomanski** [comment]
> @m-kolomanski Try looking at the following lines of code to use non-exported functions from packages:
> 
> https://github.com/Appsilon/box.linters/blob/e5658e433aa3b724650fd308fc162866f731702c/R/get_box_module_exports.R#L21-L24
> 
> Maybe, it will work. Maybe, it won't.

@radbasa Importing the functions themselves does work, however both needed functions contain call to `lintr::available_tags()` function, which fails to find the tags when `roxygen2::roxygenise()` is run (similarly to when `devtools::test()` is run).

**Comment by radbasa** [comment]
> > @m-kolomanski Try looking at the following lines of code to use non-exported functions from packages:
> > https://github.com/Appsilon/box.linters/blob/e5658e433aa3b724650fd308fc162866f731702c/R/get_box_module_exports.R#L21-L24
> > 
> > Maybe, it will work. Maybe, it won't.
> 
> @radbasa Importing the functions themselves does work, however both needed functions contain call to `lintr::available_linters()` function, which fails to find the tags when `roxygen2::roxygenise()` is run (similarly to when `devtools::test()` is run).

@jakubnowicki It would be nice if we had a list of linter tags similar to [lintr Meta-tooling](https://lintr.r-lib.org/reference/index.html#meta-tooling). But, that is automatically generated by `lintr:::rd_tags()` and `lintr:::rd_linters()` that have a lot of internal dependencies.

Do we have time to manually write documentation pages for these tags? We have 12 tags, and 15 functions.

**Comment by jakubnowicki** [comment]
> > > @m-kolomanski Try looking at the following lines of code to use non-exported functions from packages:
> > > https://github.com/Appsilon/box.linters/blob/e5658e433aa3b724650fd308fc162866f731702c/R/get_box_module_exports.R#L21-L24
> > > 
> > > Maybe, it will work. Maybe, it won't.
> > 
> > 
> > @radbasa Importing the functions themselves does work, however both needed functions contain call to `lintr::available_linters()` function, which fails to find the tags when `roxygen2::roxygenise()` is run (similarly to when `devtools::test()` is run).
> 
> @jakubnowicki It would be nice if we had a list of linter tags similar to [lintr Meta-tooling](https://lintr.r-lib.org/reference/index.html#meta-tooling). But, that is automatically generated by `lintr:::rd_tags()` and `lintr:::rd_linters()` that have a lot of internal dependencies.
> 
> Do we have time to manually write documentation pages for these tags? We have 12 tags, and 15 functions.

I would rather leave it for the next release.

**Review by radbasa — APPROVED** [review]
LGTM

**Pull request merged**